src/Mvc/Mvc.Analyzers/src/TagHelpersInCodeBlocksAnalyzer.cs

```diff
@@ -0,0 +1,157 @@
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+            context.RegisterCompilationStartAction(context =>
+            {
+                var symbolCache = new SymbolCache(context.Compilation);
+
+                if (symbolCache.TagHelperRunnerRunAsyncMethodSymbol == null)
+                {
+                    // No-op if we can't find bits we care about.
+                    return;
+                }
+
+                InitializeWorker(context, symbolCache);
+            });
+        }
+
+        private void InitializeWorker(CompilationStartAnalysisContext context, SymbolCache symbolCache)
+        {
+            context.RegisterOperationBlockStartAction(startBlockContext =>
+            {
+                startBlockContext.RegisterOperationAction(context =>
+                {
+                    var awaitOperation = (IAwaitOperation)context.Operation;
+
+                    if (awaitOperation.Operation.Kind != OperationKind.Invocation)
+                    {
+                        return;
+                    }
+
+                    var invocationOperation = (IInvocationOperation)awaitOperation.Operation;
+
+                    if (!IsTagHelperRunnerRunAsync(invocationOperation.TargetMethod, symbolCache))
+                    {
+                        return;
+                    }
+
+                    var parent = context.Operation.Parent;
+                    while (parent != null && !IsParentMethod(parent))
+                    {
+                        parent = parent.Parent;
+                    }
+
+                    if (parent == null)
+                    {
+                        return;
+                    }
+
+                    var methodSymbol = (IMethodSymbol)(parent switch
+                    {
+                        ILocalFunctionOperation localFunctionOperation => localFunctionOperation.Symbol,
+                        IAnonymousFunctionOperation anonymousFunctionOperation => anonymousFunctionOperation.Symbol,
+                        IMethodBodyOperation methodBodyOperation => startBlockContext.OwningSymbol,
+                        _ => null,
+                    });
+
+                    if (methodSymbol == null)
+                    {
+                        // Unsupported operation type.
+                        return;
+                    }
+
+                    if (!methodSymbol.IsAsync ||
+                        !symbolCache.TaskType.IsAssignableFrom(methodSymbol.ReturnType))
+                    {
+                        if (parent.Syntax.GetDiagnostics().Any(diagnostic => diagnostic.Id == TagHelperInCodeBlockDiagnostic.Id))
+                        {
+                            // Diagnostic has already been logged on this syntax item, no-op.
+                            return;
+                        }
+
+                        context.ReportDiagnostic(
```

Comment:
The typical way to de-dupe stuff btw is the following:

1. Register an operation block start operation action
2. Create a collection, or some data to track state
3. Register an operation action (use this to collect the data you need)
4. Register an operation block end operation action (use this to report diagnostics)

This way you only report a diagnostic once per block.

Reply:
> 4\. Register an operation block end operation action (use this to report diagnostics)

Do you mean a compilation end action?

Reply:
Think `RegisterOperationBlockAction` would be the corresponding end operation.